### PR TITLE
Time synchronization [5170]

### PR DESCRIPTION
**Pull request by julianbermudez**
This pull request contains the following changes:

- XRCE types have been modified according to the new version of the standard (ptc/19-03-27).
- A time synchronization mechanism has been added using the new submessages `TIEMSTAMP` and `TIMESTAMP_REPLY`.

Before review this pull request, please accept the following:

- [x] API documentation #86

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Linux/167/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Windows/127/)
* NuttX [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20NuttX/108/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Linux/168/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Windows/128/)
* NuttX [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20NuttX/109/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Linux/174/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Windows/134/)
* NuttX [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20NuttX/115/)

File: src/c/core/session/session_info.c [hunk 1]
```diff
@@ -70,15 +65,10 @@ void uxr_buffer_delete_session(uxrSessionInfo* info, ucdrBuffer* ub)
 
 void uxr_read_create_session_status(uxrSessionInfo* info, ucdrBuffer* ub)
 {
+    (void) info;
```
Comment (lemunozm) [inline src/c/core/session/session_info.c]: Remove this line

File: include/uxr/client/core/session/session.h [hunk 2]
```diff
@@ -28,10 +28,25 @@ extern "C"
...
+                                 void* args);
+
+typedef void (*uxrOnTopicFunc) (struct uxrSession* session,
+                                uxrObjectId object_id,
+                                uint16_t request_id,
+                                uxrStreamId stream_id,
+                                struct ucdrBuffer* ub,
+                                void* args);
+
+typedef void (*uxrOnTimeFunc) (struct uxrSession* session,
```
Comment (lemunozm) [inline include/uxr/client/core/session/session.h]: In what cases is this callback needed for the user?

File: test/integration/cross_serialization/ClientSerialization.cpp [hunk 3]
```diff
@@ -181,7 +171,6 @@ std::vector<uint8_t> ClientSerialization::read_data_payload()
     READ_DATA_Payload payload;
     payload.base.request_id = RequestId{0x01, 0x23};
     payload.base.object_id = ObjectId{0x45, 0x67};
-    payload.read_specification.input_stream_id = 0x80;
```
Comment (lemunozm) [inline test/integration/cross_serialization/ClientSerialization.cpp]: Did you forgot to add the `prefered_stream_id` instead?

File: src/c/core/session/session_info.c [hunk 4]
```diff
@@ -70,15 +65,10 @@ void uxr_buffer_delete_session(uxrSessionInfo* info, ucdrBuffer* ub)
...
     STATUS_AGENT_Payload payload;
     (void) uxr_deserialize_STATUS_AGENT_Payload(ub, &payload);
-
-    if(UXR_REQUEST_LOGIN == info->last_request_id)
-    {
-        uxrObjectId object_id; uint16_t request_id;
-        uxr_parse_base_object_request(&payload.base.related_request, &object_id, &request_id);
-        process_create_session_status(info, payload.base.result.status, request_id, &payload.agent_info);
```
Comment (lemunozm) [inline src/c/core/session/session_info.c]: The result status and the agent_info continuous arriving in the agent. It should be treated?

File: include/uxr/client/util/time.h [hunk 5]
```diff
@@ -25,8 +25,15 @@ extern "C"
 
 UXRDLLAPI int64_t uxr_millis(void);
 
+UXRDLLAPI int64_t uxr_nanos(void);
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply include/uxr/client/util/time.h]: It is not completely true. In a microcontroller, the `uxr_millis` return the system time but not necessarily the epoch time.

File: src/c/core/session/session_info.c (hunk 1 above)
Comment (julianbermudez) [reply src/c/core/session/session_info.c]: Fixed!

File: include/uxr/client/core/session/session.h (hunk 2 above)
Comment (julianbermudez) [reply include/uxr/client/core/session/session.h]: The user could be interested in implement her/his own time synchronization algorithm.
For example, she/he could perform a statistical analysis.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Linux/175/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Windows/135/)
* NuttX [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20NuttX/116/)

File: include/uxr/client/core/session/session.h (hunk 2 above)
Comment (lemunozm) [reply include/uxr/client/core/session/session.h]: What happened (with the stored memory of the callback) in the common case where the user does not want to implement its own sync algorithm? Should these attributes only added to the `Session` structure by a defined flag?

Comment (julianbermudez) [reply include/uxr/client/core/session/session.h]: Yes, we can add a CMake flag to enable/disable this feature. Do you think it is better to add another flag just for few bytes on the stack? Please @BorjaOuterelo, add your point of view.

Comment (BorjaOuterelo) [reply include/uxr/client/core/session/session.h]: @julianbermudez, @lemunozm  Let's go with the callback for the moment.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Linux/176/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20Windows/136/)
* NuttX [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Client%20Manual%20NuttX/117/)

**Pull request merged**